syntax = "proto3";

// Namespace
package tasks;

// Requests
message ExecuteRequest {
    // ID of the task
    bytes task_id = 1;

    // Serialized function to be executed
    bytes function = 2;

    // List of function arguments
    bytes args = 3;
}

message CancelRequest {
    // ID of the task
    bytes task_id = 1;
}

message ScheduleTaskRequest {
    // ID of the task
    bytes task_id = 1;
}

message GetResourceLoadRequest {
}


// Responses
message ExecuteReply {
    // ID of the task
    bytes task_id = 1;

    // Return value of task
    bytes result = 2;
}

message CancelReply {
    // ID of the task
    bytes task_id = 1;
}

message ScheduleTaskReply {
    // ID of the task
    bytes task_id = 1;
}

message GetResourceLoadResponse {
    float cpu_load;
    int64 memory_used;
}

// Service responsible for tasks
service TaskService {
    // Executes a task
    rpc ExecuteTask (ExecuteRequest) returns (ExecuteReply);

    // Add a task to scheduler
    rpc AddTaskToScheduler (ScheduleTaskRequest) returns (ScheduleTaskReply);
    
    // Cancels running task
    rpc CancelTask (CancelRequest) returns (CancelReply);

    // Get Resource Load From Server
    rpc GetResourceLoad (GetResourceLoadRequest) returns (GetResourceLoadResponse);
}